samza-sql/src/main/java/org/apache/samza/sql/data/RexToJavaCompiler.java

```diff
@@ -0,0 +1,241 @@
+    final JavaTypeFactoryImpl javaTypeFactory = new JavaTypeFactoryImpl(rexBuilder.getTypeFactory().getTypeSystem());
+
+    // public void execute(Object[] inputValues, Object[] outputValues)
+    final RexToLixTranslator.InputGetter inputGetter = new RexToLixTranslator.InputGetterImpl(ImmutableList.of(
+        Pair.<org.apache.calcite.linq4j.tree.Expression, PhysType>of(
+            Expressions.variable(Object[].class, "inputValues"),
+            PhysTypeImpl.of(javaTypeFactory, inputRowType, JavaRowFormat.ARRAY, false))));
+    final Function1<String, RexToLixTranslator.InputGetter> correlates = a0 -> {
+      throw new UnsupportedOperationException();
+    };
+    final List<org.apache.calcite.linq4j.tree.Expression> list =
+        RexToLixTranslator.translateProjects(program, javaTypeFactory, builder, null, null, inputGetter, correlates);
+    for (int i = 0; i < list.size(); i++) {
+      builder.add(Expressions.statement(
+          Expressions.assign(Expressions.arrayIndex(outputValues, Expressions.constant(i)), list.get(i))));
+    }
+    return createSamzaExpressionFromCalcite(inputValues, outputValues, builder.toBlock());
+  }
+
+  /**
+   * Given a method that implements {@link org.apache.samza.sql.data.Expression#execute(Object[], Object[])},
+   * adds a bridge method that implements {@link org.apache.samza.sql.data.Expression#execute(Object[])}, and
+   * compiles.
+   */
+  static org.apache.samza.sql.data.Expression createSamzaExpressionFromCalcite(ParameterExpression inputValues,
+      ParameterExpression outputValues, BlockStatement block) {
+    final List<MemberDeclaration> declarations = Lists.newArrayList();
+
+    // public void execute(Object[] inputValues, Object[] outputValues)
+    declarations.add(
+        Expressions.methodDecl(Modifier.PUBLIC, void.class, SamzaBuiltInMethod.EXPR_EXECUTE2.method.getName(),
+            ImmutableList.of(inputValues, outputValues), block));
+
+    // public Object execute(Object[] inputValues)
+    final BlockBuilder builder = getExecuteMethodWithOnlyInputsBlockBuilder(inputValues);
+
+    declarations.add(
+        Expressions.methodDecl(Modifier.PUBLIC, Object.class, SamzaBuiltInMethod.EXPR_EXECUTE1.method.getName(),
+            ImmutableList.of(inputValues), builder.toBlock()));
+
+    final ClassDeclaration classDeclaration = Expressions.classDecl(Modifier.PUBLIC, "SqlExpression", null,
+        ImmutableList.<Type>of(org.apache.samza.sql.data.Expression.class), declarations);
+    String s = Expressions.toString(declarations, "\n", false);
+
+    if (log.isDebugEnabled()) {
+      log.debug("Generated code for expression: " + s);
+    }
+
+    try {
+      return getExpression(classDeclaration, s);
+    } catch (Exception e) {
+      throw new SamzaException("Expression compilation failure.", e);
+    }
+  }
+
+  static BlockBuilder getExecuteMethodWithOnlyInputsBlockBuilder(ParameterExpression inputValues) {
```

Comment:
Can we add some documentation/example to illustrate what this method is implementing? It is pretty hard to follow this code w/o browsing into the Calcite code. It would be helpful to add some simple SQL example and show how each of this method is used to implement which part of SQL statement.